pkg/token/serviceprincipaltoken.go

```diff
@@ -51,32 +58,55 @@ func newServicePrincipalToken(oAuthConfig adal.OAuthConfig, clientID, clientSecr
 		clientCertPassword: clientCertPassword,
 		resourceID:         resourceID,
 		tenantID:           tenantID,
-		oAuthConfig:        oAuthConfig,
+		cloud:              *cloud,
 	}, nil
 }
 
+// Token fetches an azcore.AccessToken from the Azure SDK and converts it to an adal.Token for use with kubelogin.
 func (p *servicePrincipalToken) Token() (adal.Token, error) {
-	emptyToken := adal.Token{}
-	callback := func(t adal.Token) error {
-		return nil
-	}
+	return p.TokenWithOptions(nil)
+}
 
-	var (
-		spt *adal.ServicePrincipalToken
-		err error
-	)
+func (p *servicePrincipalToken) TokenWithOptions(options *azcore.ClientOptions) (adal.Token, error) {
+	emptyToken := adal.Token{}
+	var spnAccessToken azcore.AccessToken
 
+	// Request a new Azure token provider for service principal
 	if p.clientSecret != "" {
-		spt, err = adal.NewServicePrincipalToken(
-			p.oAuthConfig,
+		clientOptions := &azidentity.ClientSecretCredentialOptions{
+			ClientOptions: azcore.ClientOptions{
+				Cloud: p.cloud,
+			},
+		}
+		if options != nil {
+			clientOptions.ClientOptions = *options
+		}
+		cred, err := azidentity.NewClientSecretCredential(
+			p.tenantID,
 			p.clientID,
 			p.clientSecret,
-			p.resourceID,
-			callback)
+			clientOptions,
+		)
+		if err != nil {
+			return emptyToken, fmt.Errorf("unable to create credential. Received: %v", err)
```

Comment:
generally wrapping error is preferred. you will see that using %v is even banned in our team :)
https://www.google.com/search?q=golang+error+wrapping

Reply:
Fixed